src/charm.py

```diff
@@ -52,18 +52,39 @@ def __init__(self, *args: Any) -> None:
             host=f"{self.app.name}-endpoints.{self.model.name}.svc.cluster.local",
             strip_prefix=True,
         )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.reset_instance_action, self._reset_instance_action)
 
     def _on_config_changed(self, event: ops.HookEvent) -> None:
         """Handle changed configuration.
 
         Args:
             event: Event triggering after config is changed.
         """
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
         container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
         if not container.can_connect():
             event.defer()
             self.unit.status = ops.WaitingStatus("Waiting for pebble")
             return
+        server_name_configured = self._synapse.server_name_configured(container)
+        if (
+            server_name_configured is not None
+            and server_name_configured != self._charm_state.server_name
+        ):
+            msg = (
+                f"server_name {self._charm_state.server_name} is different from the existing one"
+                f" {server_name_configured}."
+                " Please revert the config or"
+                " run the action reset-instance if you to erase the existing instance and start a"
+                " new one."
+            )
+            self.model.unit.status = ops.BlockedStatus(msg)
```

Comment:
Minor nitpick, I'm wondering if this message is too lengthy to be set as status message. Would logging be better and the status message direct the users to the logs?

Reply:
Good catch, changed :)